custom_components/spotcast/__init__.py

```diff
@@ -183,11 +184,10 @@ def getSpotifyConnectDeviceId(client, device_name):
             spotify_cast_device.startSpotifyController(access_token, expires)
             spotify_device_id = spotify_cast_device.getSpotifyDeviceId(client)
 
-        transfer_playback = shouldTransferPlayback(call, client)
-        if transfer_playback == True:
+        if shouldTransferPlayback(call, client) == True:
             _LOGGER.debug('Transfering playback')
             current_playback = client.current_playback()
-            if current_playback is not None:
+            if current_playback is not None or transfer_playback == True:
```

Comment:
`shouldTransferPlayback` has the following check:
```
 if uri is None or uri.strip() == '' or call.data.get(CONF_TRANSFER_PLAYBACK):
````

I'm wondering what scenario would cause line 190 to also be false so that line 194 to run. Perhaps it will make more sense in this PR to remove the function `shouldTransferPlayback` alltogether to make it clear what the flow is.

Reply:
Hi Niklas,

The scenario I had in mind is:
- Spotify is not playing
- User is calling the start service without uri and without transfer playback (or transfer playback = false)

In this scenario the device will "turn on" but not continue playing where the user last stopped (=run line 194).

If...
- Spotify is not playing
- User is calling the start service without uri and with transfer playback = true
...the device will "turn on" and continue playing where the user last stopped (=run line 192).

To complex? :-)

Best
 Christian

Reply:
Maybe not too complex but I would love to be able to follow the logic in a year ;-)

I guess listing the use cases that needs to be supported might help.

1. Spotify is playing and user wants to change device for the current playback
2. Spotify is not playing and the user wants to start spotify on the device with whatever was playing last
3. A variant of #1 where the user wants to start or continue playback on a particular device regardless of if spotify is playing something or not.
4. User wants to start spotify in a device but don't want to play anything (currently not supported and I doubt I like to support it)

Any other use cases?

Reply:
Can't think of any more use cases.

I use use case 4 in my morning routine:
- Start Spotify in a device, but do not start playing
- Turn down volume to a very low value
- Start playing on that device
- Step by step turn volume up

I achieve use case 4 by calling start with no uri and transfer playback = false (using the code from the pull request).

Reply:
I think there are others that want to avoid the bloink sound. Maybe better to implement that natively with setting the volume to 0 at the start of `startSpotifyController` and raising it again to the old level (first checking for volume setting).

like so:
```
    def startSpotifyController(self, access_token, expires):
        from pychromecast.controllers.spotify import SpotifyController
        # get the volume so we can remove the bloink
        volume = self.castDevice.status.volume_level
        self.castDevice.set_volume(0)

        sp = SpotifyController(access_token, expires)
        self.castDevice.register_handler(sp)
        sp.launch_app()
        
        # reset the volume
        self.castDevice.set_volume(volume)

        if not sp.is_launched and not sp.credential_error:
            raise HomeAssistantError('Failed to launch spotify controller due to timeout')
        if not sp.is_launched and sp.credential_error:
            raise HomeAssistantError('Failed to launch spotify controller due to credentials error')

        self.spotifyController = sp
```

Reply:
But if you want to very slowly raise the volume you could lower the volume before spotcast.start and then use the media_player entity to raise the volume slowly, or?

Reply:
I agree.
The sketched way won't work with Spotify Connect devices though, but that's not what Spotcast is for anyways. :-)

Reply:
> `shouldTransferPlayback` has the following check:
> 
> ```
>  if uri is None or uri.strip() == '' or call.data.get(CONF_TRANSFER_PLAYBACK):
> ```
> 
> I'm wondering what scenario would cause line 190 to also be false so that line 194 to run. Perhaps it will make more sense in this PR to remove the function `shouldTransferPlayback` alltogether to make it clear what the flow is.

It could look like that:
```
        if uri is None or uri.strip() == '':
            _LOGGER.debug('Transfering playback')
            current_playback = client.current_playback()
            if current_playback is not None or transfer_playback == True:
                _LOGGER.debug('current_playback from spotipy: %s', current_playback)
                client.transfer_playback(device_id=spotify_device_id, force_play=True)
            else:
                client.transfer_playback(device_id=spotify_device_id, force_play=False)
        else:
            play(client, spotify_device_id, uri, random_song, repeat, shuffle, position)
```
However, the parameter name should rather be "force_playback" or similar; plus documentation needs an update as the way to actually just transfer the playback is caling start with no uri.

Reply:
I'm fine with changing to that. I have a slight recollection of the force having problems in some cases though so this new behaviour needs documentation. Will you update the PR?